### ISSUE TITLE
Cast exception when reading DateTimeOffset with System.Data.SqlClient

### ISSUE THREAD

**Issue by daniellittledev** [pimbrouwers/Donald#28]
SqlClient supports a native DateTimeOffset type but it looks like Donald assumes the value will be stored as a string. The following line shows how it is currently working.

https://github.com/pimbrouwers/Donald/blob/8592f096c50a0a23e48f36b62017e67c00e2d3f5/src/Donald/DataReader.fs#L56

Therefore when using the data type `DateTimeOffset` you get the following exception:

```
Exception thrown: 'System.InvalidCastException' in System.Data.SqlClient.dll
Unable to cast object of type 'System.DateTimeOffset' to type 'System.String'.
```

What's the best way to read a `DateTimeOffset` with Donald, would we need a database specific assembly to read database specific types?

At the very least it might be worth removing `ReadDateTimeOffset` if it causes exceptions depending on the database.

**Comment by pimbrouwers** [comment]
Hey Daniel! Thanks for bringing this to my attention, admittedly it's not a field type I use or encounter very often. So it's no suprise that I didn't catch this myself (since I use this package everyday at work with SQL Server) and if you look carefully at the "type reader test" the mapper is missing the `datetimeoffset` 🤦.

So let's solve the two problems in front of us. First and most importantly, getting your code working. Here's a solution for right now.

```fsharp
open System.Data.SqlClient

// quick & dirty
let mapRow (rd : IDataReader) : DateTimeOffset =
    let sqlRd = rd :?> SqlDataReader
    sqlRd.GetDateTimeOffset(sqlRd.GetOrdinal("dto"))

conn
|> Db.newCommand "SELECT SYSDATETIMEOFFSET() AS dto"
|> Db.query mapRow

// longer term
module Db =
    let queryMsSql (map : SqlDataReader -> 'a) (cmd : SqlCommand) : DbResult<'a list> =   
        let map' (r : IDataReader) = r :?> SqlDataReader |> map
        let cmd' = cmd :> IDbCommand
        Db.query map' cmd'

let mapRow (rd : SqlDataReader) : DateTimeOffset =
    rd.GetDateTimeOffset(sqlRd.GetOrdinal("dto"))

conn
|> Db.newCommand "SELECT SYSDATETIMEOFFSET() AS dto"
|> Db.queryMsSql mapRow
```

Now as far as the package goes, I think I have a couple ideas. Regardless of how we proceed on that, first thing should be to mark the current `ReadDateTimeOffset` as obsolete, release as a patch and remove it in the next major version release. 

In my view the cleanest solution is to apply generics. If we change the signature of the map param to the following, you can pass in any function which receives any implementation of the interface.

```fsharp
let query (map : 'reader -> 'a when 'reader :> IDataReader) (cmd : IDbCommand) : DbResult<'a list> =  // ...

// now this works
let mapRow (rd : SqlDataReader) : DateTimeOffset =
    rd.GetDateTimeOffset(sqlRd.GetOrdinal("dto"))

conn
|> Db.newCommand "SELECT SYSDATETIMEOFFSET() AS dto"
|> Db.query mapRow
```

At first glance, I think this idea is our best approach. It's regression-proof and immediately, without any work allows people to plugin vendor specific consumption.

Thoughts?

**Comment by daniellittledev** [comment]
Hey, thanks for quick fix and all the detail!

Yeah I think that looks like a great path forward. Switching query to use generics provides easy accesses to the  
more specific IDataReader type and looks like it's backwards compatible too!

**Comment by pimbrouwers** [comment]
Awesome! This has been merged in and deployed through 6.2.3

**Comment by pimbrouwers** [comment]
Ref: commit 41a7b6cfd408ed70be1eeb2fe0468392c8a608b6
